_posts/2024-06-19-optimizing-opensearch-with-fp16-quantization.md

```diff
@@ -0,0 +1,288 @@
+
+`1.1 * (2 * 256 + 8 * 16) * 1,000,000 ~= 0.656 GB`
+
+For more information about memory estimation for scalar quantization with the inverted file (IVF) algorithm, refer to [this documentation](https://opensearch.org/docs/latest/search-plugins/knn/knn-vector-quantization/#memory-estimation-1).
+
+## Benchmarking results
+
+We ran benchmarking tests on some popular and trending datasets using our [opensearch-benchmark](https://github.com/opensearch-project/opensearch-benchmark-workloads/tree/main/vectorsearch) tool 
+to compare the indexing, search performance, and quality of search results of Faiss scalar quantization. We compared Faiss scalar quantization against using Faiss with float vectors without any encoding. All tests were performed with [SIMD](https://opensearch.org/docs/latest/search-plugins/knn/knn-index/#simd-optimization-for-the-faiss-engine) (Single Instruction Multiple Data).
+enabled on x86 architecture with AVX2 optimization.
+
+**Note**: Without SIMD optimization (AVX2 or NEON) or with AVX2 disabled (on x86 architecture), the quantization process introduces additional overhead, which leads to an increase in latency. 
+For information about processors that support AVX2, see [CPUs with AVX2](https://en.wikipedia.org/wiki/Advanced_Vector_Extensions#CPUs_with_AVX2). In AWS, all community Amazon Machine Images (AMIs) with [HVM](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/virtualization_types.html) support AVX2 optimization for the x86 architecture.
+
+### Benchmarking results using small workloads
+
+We ran the following tests on a single-node cluster without any replicas using the following datasets.
+
+
+#### Configuration
+
+|m	|ef_construction	|ef_search	|replica|
+|---	|---	|---	|---	|
+|16	|100	|100	|0	|
+
+The dataset and other configuration details are listed in the following table.
+
+|Dataset ID	|Dataset	|Vector dimension	|Data size	|Number of queries	|Training data range	|Query data range	|Space type	|Primary shards	|Indexing clients|
+|---	|---	|---	|---	|---	|---	|---	|---	|---	|---	|
+|Dataset 1	|gist-960-euclidean	|960	|1,000,000	|1,000	|[ 0.0, 1.48 ]	|[ 0.0, 0.729 ]	|L2	|8	|16|
+|Dataset 2	|mnist-784-euclidean	|784	|60,000	|10,000	|[ 0.0, 255.0 ]	|[ 0.0, 255.0 ]	|L2	|1	|2|
+|Dataset 3	|cohere-wiki-simple-embeddings-768	|768	|475,858	|10,000	|[ -4.1561704, 5.5478516 ]	|[ -4.065383, 5.4902344 ]	|L2	|4	|8|
+|Dataset 4	|cohere-ip-1m	|768	|1,000,000	|10,000	|[ -4.1073565, 5.504557 ]	|[ -4.109505, 5.4809895 ]	|innerproduct	|8	|16|
+|Dataset 5	|sift-128-euclidean	|128	|1,000,000	|10,000	|[ 0.0, 218.0 ]	|[ 0.0, 184.0 ]	|L2	|8	|16|
+
+#### Recall and memory results
+
+|Dataset ID	| Faiss hnsw recall@100	|Faiss hnsw-sqfp16 recall@100	|Faiss hnsw memory estimate (gb)	|Faiss hnsw-sqfp16 memory estimate (gb)	|Faiss hnsw memory usage (gb)	|Faiss hnsw-sqfp16 memory usage (gb)	|% reduction in memory	|
+|---	|---	|---	|---	|---	|---	|---	|---	|
```

Comment:
Are four digits really meaningful for recall? I'd think two would be enough (0.91, not 0.9071).

Reply:
If we round it to two digits, users might think that the recall would be same with fp32 or fp16 but it is actually not as there would be a slight drop in recall with fp16. So, want to show that difference to make it more transparent

Reply:
In Dataset 1, recall actually _increases_ by 0.0001. That is clearly not meaningful.
And I would claim that all the other differences are so small that they are not meaningful.
The lesson is not that we need to report four digits, but that we need to report cases which show a meaningful difference. For example, what is recall @10? (Too late for this doc, but you get the idea.)

Reply:
For dataset1, that could be due to some noise which I believe can be ignored. If you feel that it is negligible and can be ignored, I will round them to 2 digits.

recall@10 signifies recall measuring at k=10. Unfortunately, we don't have metrics for other values of k to show those tradeoffs. But, will make a note of it and will add those in the future blogs.